### ISSUE TITLE
Prediction is dependent on other predicted data points

### ISSUE THREAD

**Issue by rachtsingh** [cornellius-gp/gpytorch#1704]
# 🐛 Bug

This might not be a bug but instead my misunderstanding of how GPyTorch works, but my understanding is that conditional on the rest of the data points in the GP, the prediction for any given data point is independent of what we ask about other data points.  However, I see that asking for certain data points in prediction is adding some numerical instability, which I assume is due to some optimization that is numerically unstable?

## To reproduce

Unfortunately, I don't have a reproducer I can share, but I can clearly describe the data and model:

model: 
```
ExactGPModel(
  (likelihood): FixedNoiseGaussianLikelihood(
    (noise_covar): FixedGaussianNoise()
  )
  (mean_module): ConstantMean()
  (covar_module): ScaleKernel(
    (base_kernel): RBFKernel(
      (lengthscale_prior): NormalPrior()
      (raw_lengthscale_constraint): GreaterThan(5.000E+01)
      (distance_module): Distance()
    )
    (raw_outputscale_constraint): Positive()
  )
  (feature_extractor): CustomLengthScaleExtractor()
)
```
where `CustomLengthScaleExtractor` is a piece-wise monotonic function that just transforms x before it passes through the GP (I don't think it's super relevant to the problem, but I can explain more). Essentially this is a 1-dimensional GP with some data points `(x, y)` such that all the data has `x` in `[-20, 540]` and `y` between `0, 12`. The FixedGaussianNoise is pretty much all set to 1, with some data points having noise of 3.

** Stack trace/error message **
This is if we ask only for data points inside the range:

```python
xpred = torch.arange(-20, 500).to(torch.float32)
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    # observed_pred = likelihood(model(test_x))
    f_preds = gpmodel(xpred)
    f_mean = f_preds.mean

mean = f_preds.mean.detach().cpu().numpy()
std = f_preds.stddev.detach().cpu().numpy()
plt.figure(figsize=(8, 6))
plt.plot(xpred.numpy(), mean, '-', color='gray')

plt.fill_between(xpred.numpy(), mean - std, mean + std,
                 color='gray', alpha=0.2)
plt.xlim(-5, 5)
plt.ylim(10, 11)
```
![ok](https://user-images.githubusercontent.com/1606892/126354695-303ee0aa-7445-4595-815d-82653773669c.png)

However when I expand the range, I get some noisy predictions:

```
xpred = torch.arange(-20, 1000).to(torch.float32)
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    # observed_pred = likelihood(model(test_x))
    f_preds = gpmodel(xpred)
    f_mean = f_preds.mean

mean = f_preds.mean.detach().cpu().numpy()
std = f_preds.stddev.detach().cpu().numpy()
plt.figure(figsize=(8, 6))
plt.plot(xpred.numpy(), mean, '-', color='gray')

plt.fill_between(xpred.numpy(), mean - std, mean + std,
                 color='gray', alpha=0.2)
plt.xlim(-5, 5)
plt.ylim(10, 11)
plt.tight_layout()
```

![bad](https://user-images.githubusercontent.com/1606892/126357120-302e763d-baf6-4c2a-864e-dd37aef5d4ef.png)

I also get the following warning (which I guess might be because of `fast_pred_var`):
```
..site-packages/gpytorch/distributions/multivariate_normal.py:263: NumericalWarning: Negative variance values detected. This is likely due to numerical instabilities. Rounding negative variances up to 1e-06.
  NumericalWarning,
```

## Expected Behavior

I expected that as we asked for a larger range of predictions (including the earlier values), the predictions on the earlier set shouldn't change. In other words, if I predict on A, then predict on A, B, then the two predictions for A  (definitely in mean and probably in variance?) should be the same. Please let me know if I just misunderstand!

## System information

**Please complete the following information:**
- GPyTorch version `1.4.1`
- PyTorch version `1.7.1+cu101`
- OS is RHEL

**Comment by Balandat** [comment]
This is probably due to the accuracy of the solve. Note that when you compute the prediction for the larger range you're also computing it on a lot more data points (1020 vs 520). My hunch is that that internally the solve of the linear system is switching from using a Cholesky decomposition to using a Linear Conjugate Gradient solve, which is less accurate, hence the difference. Try running the larger prediction with a `with gpytorch.settings.max_cholesky_size(2000):` context manager and see what happens. 

Aside: Note also that the way you're computing things you will get the joint distribution over the data points. If you don't need the covariances you can batch-evaluate the model (use `test_x.unsqueeze(-1)`) to only compute marginal variances.

**Comment by rachtsingh** [comment]
Hmm, that doesn't seem to be it. I tried this:

```python
xpred = torch.arange(-20, 1000).to(torch.float32)
with torch.no_grad(), gpytorch.settings.fast_pred_var(), gpytorch.settings.max_cholesky_size(2000):
    f_preds = gpmodel(xpred)
    f_mean = f_preds.mean

mean = f_preds.mean.detach().cpu().numpy()
std = f_preds.stddev.detach().cpu().numpy()
plt.figure(figsize=(8, 6))
plt.plot(xpred.numpy(), mean, '-', color='gray', label='xpred=[-20, {:.0f}]'.format(xpred.max()))
plt.plot(xpred1.numpy(), mean1, '-', color='green', label='xpred=[-20, {:.0f}]'.format(xpred1.max()))

plt.fill_between(xpred.numpy(), mean - std, mean + std, color='gray', alpha=0.2)

plt.xlim(-5, 5)
plt.ylim(10, 11)
plt.legend()
plt.tight_layout()
plt.grid(ls='--')
```

and get this: 

![bad4](https://user-images.githubusercontent.com/1606892/126402816-50781d61-a00b-45bf-afb2-536f137de618.png)

(and still getting the numerical warning, fwiw).

Thanks for the help! The comment about not computing the joint distribution is very helpful, thank you.

**Comment by rachtsingh** [comment]
Hmm, I must be doing something wrong. I just tried this (on a different, but similarly structured model):

```python
print(model(torch.tensor([-20, 0.]).cuda()).mean)
print(model(torch.tensor([0.]).cuda()).mean)
```

and got:

```
tensor([ 9.7718, 10.5228], device='cuda:0', grad_fn=<ViewBackward>)
tensor([11.5386], device='cuda:0', grad_fn=<ViewBackward>)
```

**Comment by rachtsingh** [comment]
I've managed to reproduce the weird behavior above with a dataset that I can share (attached). 

Here's the full code for an example:

```python
import gpytorch
import numpy as np
import torch
from matplotlib import pyplot as plt

# load this data wherever you've saved it
x = np.loadtxt("x.csv", delimiter=",")
y = np.loadtxt("y.csv", delimiter=",")

# copied from the KISS-GP tutorial
train_x = torch.tensor(x)
train_y = torch.tensor(y)

class GPRegressionModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)

        # SKI requires a grid size hyperparameter. This util can help with that. Here we are using a grid that has the same number of points as the training data (a ratio of 1.0).
        # Performance can be sensitive to this parameter, so you may want to adjust it for your own problem on a validation set.
        grid_size = gpytorch.utils.grid.choose_grid_size(train_x, 1.0)

        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.GridInterpolationKernel(
                gpytorch.kernels.RBFKernel(), grid_size=grid_size, num_dims=1
            )
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = GPRegressionModel(train_x, train_y, likelihood)

model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam([
    {'params': model.parameters()},  # Includes GaussianLikelihood parameters
], lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

for i in range(30):
    optimizer.zero_grad()
    output = model(train_x)
    loss = -mll(output, train_y)
    print("{:.3f}".format(loss.item()))
    loss.backward()
    optimizer.step()

model.eval()
likelihood.eval()

# Initalize plot
f, ax = plt.subplots(1, 1, figsize=(8, 6))

# The gpytorch.settings.fast_pred_var flag activates LOVE (for fast variances)
# See https://arxiv.org/abs/1803.06058
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(-20, 100, 200)
    prediction = likelihood(model(test_x))
    mean = prediction.mean
    # Get lower and upper predictive bounds
    lower, upper = prediction.confidence_region()
    
# Plot the training data as black stars
def ax_plot():
    ax.plot(train_x.detach().numpy(), train_y.detach().numpy(), 'k*', alpha=0.2)
    # Plot predictive means as blue line
    ax.plot(test_x.detach().numpy(), mean.detach().numpy(), 'b')
    # Plot confidence bounds as lightly shaded region
    ax.fill_between(test_x.detach().numpy(), lower.detach().numpy(), upper.detach().numpy(), alpha=0.5)
    ax.set_xlim([-20, 100])
    ax.grid(ls='--', color="#ddd")
    ax.legend(['Observed Data', 'Mean', 'Confidence'])

ax_plot()
```

The output I see is this:

![bad5](https://user-images.githubusercontent.com/1606892/126530925-dfebe92a-eb0b-40be-80b6-aa33477df401.png)

So it seems like there's some weird offsetting behavior I'm getting, which doesn't make sense to me at first glance. 

But then running this also results in weird responses:

```python
print(model(torch.tensor([-20, -10., -5., 0.]).unsqueeze(-1)).mean)
print(model(torch.tensor([0.]).unsqueeze(-1)).mean)
print(model(torch.tensor([0., 10.]).unsqueeze(-1)).mean)
```

```
tensor([ 9.7762, 11.5278, 10.9341, 10.5967], grad_fn=<ViewBackward>)
tensor([11.5530], grad_fn=<ViewBackward>)
tensor([11.5530, 10.6881], grad_fn=<ViewBackward>)
```

I actually think these issues are linked. The first prediction for 0 (first line) is from the blue line; whereas the second and third lines have the "real" prediction (i.e. the center of the marginal distribution at x=0 as observed from the black marks).

**I think this particular problem is fixed by manually setting grid bounds in GridInterpolationKernel**. Can I suggest that the default grid bounds be determined by train_x.min() and train_x.max()? It seems user-friendly. I don't quite understand how that could cause the above issue but I guess I should read the paper. 

I'm not sure this solves the jittery prediction issue above, but I'll have to try again.

**Comment by wjmaddox** [comment]
Hi, sorry I'm not able to access the data now that I'm taking a look at this, but will try simulating some data to look roughly similar. 

A priori, I'd expect that part of what's happening is that the grid is changing (e.g. expanding if you predict at certain data points) by virtue of this set of code: https://github.com/cornellius-gp/gpytorch/blob/c074c2ff5ba5708761453bbd9be870c35cb57769/gpytorch/kernels/grid_interpolation_kernel.py#L158.  In general, this should be nearly equivalent to setting the default grid bounds to be `train_x.min()` and `train_x.max()`, but on a dimension-wise setting.

**Comment by wjmaddox** [comment]
Anyways, I'm pretty confident that what you're seeing as the predictions changing is really the SKI grid changing in response to seeing values of `x` that are outside the range of the data. 

```python
x = torch.rand(8000, 1) * 120 - 10
y = torch.exp(-x / 30) + torch.exp(-torch.sin((x  - 3.5) / 1.5) / 10) + 10 + 0.25 * torch.randn_like(x)
# this data sort of looks like yours, but is crucially on the same scale [-10, 110]

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = GPRegressionModel(train_x, train_y, likelihood)

model.train()
likelihood.train()

# now print the grid
print(model.covar_module.base_kernel.grid[0])
# tensor([ -9.9977,  -9.9827,  -9.9677,  ..., 110.0126, 110.0276, 110.0426])

# The gpytorch.settings.fast_pred_var flag activates LOVE (for fast variances)
# See https://arxiv.org/abs/1803.06058
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(-20, 100, 200).double()
    prediction = likelihood(model(test_x))
    mean = prediction.mean
    # Get lower and upper predictive bounds
    lower, upper = prediction.confidence_region()

# check the grid
print(model.covar_module.base_kernel.grid[0])
# tensor([ -9.9977,  -9.9827,  -9.9677,  ..., 110.0126, 110.0276, 110.0426])

# it also changes if you put the print statements in between your comments
print(model.covar_module.base_kernel.grid[0])
print("-20: ", model(torch.tensor([-20, -10., -5., 0.]).unsqueeze(-1).double()).mean)
print(model.covar_module.base_kernel.grid[0])
print("0", model(torch.tensor([0.]).unsqueeze(-1).double()).mean)
print(model.covar_module.base_kernel.grid[0])
print("0, 10", model(torch.tensor([0., 10.]).double().unsqueeze(-1)).mean)
print(model.covar_module.base_kernel.grid[0])
####
# tensor([ -9.9977,  -9.9827,  -9.9677,  ..., 110.0126, 110.0276, 110.0426],
 #      dtype=torch.float64)
# -20:  tensor([12.3671, 11.8522, 11.8437, 11.7605], dtype=torch.float64,
 #      grad_fn=<ViewBackward>)
# tensor([-20.0489, -20.0327, -20.0164,  ..., 110.0138, 110.0301, 110.0464],
 #      dtype=torch.float64)
# 0 tensor([11.9508], dtype=torch.float64, grad_fn=<ViewBackward>)
# tensor([ -9.9977,  -9.9827,  -9.9677,  ..., 110.0126, 110.0276, 110.0426],
#       dtype=torch.float64)
# 0, 10 tensor([11.9508, 11.9698], dtype=torch.float64, grad_fn=<ViewBackward>)
# tensor([ -9.9977,  -9.9827,  -9.9677,  ..., 110.0126, 110.0276, 110.0426],
#       dtype=torch.float64)
```

The prediction changes are probably most noticeable because your data has wide range and isn't standardized to say [0,1] or to have zero mean.
